pandas/_libs/tslibs/conversion.pyx

```diff
@@ -347,9 +349,14 @@ cdef _TSObject convert_datetime_to_tsobject(datetime ts, object tz,
     if tz is not None:
         tz = maybe_get_tz(tz)
 
-        # sort of a temporary hack
         if ts.tzinfo is not None:
             if hasattr(tz, 'normalize') and hasattr(ts.tzinfo, '_utcoffset'):
+                # tz.localize does not correctly localize Timestamps near DST
+                # but correctly localizes datetimes
+                if hasattr(ts, 'to_pydatetime'):
```

Comment:
Elsewhere we've used `not PyDatetimeCheckExact(ts)` as a standin for `isinstance(ts, pd.Timestamp)`.  This might also be use case for the `nanos` kwarg.

Reply:
Thanks @jbrockmendel. Seems clearer than trying to get `Timestamp` to quack